[feature](nereids) Plan Translator

Kikyou1997 opened the pull request:
# Proposed changes

Issue Number: close #9621

## Problem Summary:

Add following physical operator: PhysicalAgg PhysicalSort PhysicalHashJoin

Add basic logic of plan translator

## Checklist(Required)

1. Does it affect the original behavior: (No)
2. Has unit tests been added: (No)
3. Has document been added or modified: (No)
4. Does it need to update dependencies: (No)
5. Are there any changes that cannot be rolled back: (No)

Kikyou1997 committed:
plan translator

Kikyou1997 committed:
Merge branch 'master' of github.com:apache/incubator-doris into upstream_master

Kikyou1997 committed:
Merge branch 'master' of github.com:apache/incubator-doris into opt

Kikyou1997 committed:
add PhysicalSort PhysicalAgg PhysicalHashJoin and some basic logic of plan translator

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/PhysicalPlanTranslator.java:
```diff
@@ -0,0 +1,309 @@
+                }
+                break;
+            case FIRST_MERGE:
+                aggInfo = AggregateInfo.create(execGroupingExpressions, execAggExpressions, outputTupleDesc,
```
do we need exchange node under agg merge node?

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/planner/ExchangeNode.java:
```diff
@@ -64,6 +64,8 @@ public class ExchangeNode extends PlanNode {
     // only if mergeInfo_ is non-null, i.e. this is a merging exchange node.
     private long offset;
 
+    private DataPartition dataPartition;
```
why add this?

Kikyou1997 replied:
removed

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalSort.java:
```diff
@@ -0,0 +1,75 @@
+    private List<OrderKey> orderList;
+
+    // if true, the output of this node feeds an AnalyticNode
+    private boolean isAnalyticSort;
```
current, we do not support analytic functions. In old planner, this is a trick flag. In Nereids, this should no longer exists

Kikyou1997 replied:
removed

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalOperator.java:
```diff
@@ -29,4 +29,6 @@
  */
 public interface PhysicalOperator<TYPE extends PhysicalOperator<TYPE>> extends PlanOperator<TYPE> {
     List<Slot> computeOutputs(LogicalProperties logicalProperties, Plan... inputs);
+
```
blank lines

Kikyou1997 replied:
removed

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalOlapScan.java:
```diff
@@ -34,6 +35,8 @@ public class PhysicalOlapScan extends PhysicalScan<PhysicalOlapScan> {
     private final List<Long> selectedTabletId;
     private final List<Long> selectedPartitionId;
 
+    private OlapTable olapTable;
```
why move this from parent class to sub class

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalAggregation.java:
```diff
@@ -0,0 +1,82 @@
+
+    private List<Expression> partitionExprList;
+
+    private AggregateInfo.AggPhase aggPhase;
```
maybe a new enum is better

Kikyou1997 replied:
done

morrySnow commented:
fix check style~

Kikyou1997 commented on fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/PhysicalPlanTranslator.java:
```diff
@@ -0,0 +1,309 @@
+                }
+                break;
+            case FIRST_MERGE:
+                aggInfo = AggregateInfo.create(execGroupingExpressions, execAggExpressions, outputTupleDesc,
```
exactly

Kikyou1997 commented on fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/Plan.java:
```diff
@@ -44,4 +44,6 @@
 
     @Override
     Plan child(int index);
+
```
removed

Kikyou1997 committed:
fix code style

Kikyou1997 committed:
1. add new agg phase enum for nereids
2. remove the Analyzer from PlanContext.java
3. implement PlanTranslator::visitPhysicalFilter

Kikyou1997 committed:
merge master

Kikyou1997 commented on fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/PlanContext.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    public TupleDescriptor generateTupleDesc() {
+        return analyzer.getDescTbl().createTupleDescriptor();
```
done

jackwener commented on fe/fe-core/src/main/java/org/apache/doris/planner/HashJoinNode.java:
```diff
@@ -140,6 +141,60 @@ public HashJoinNode(PlanNodeId id, PlanNode outer, PlanNode inner, TableRef inne
         }
     }
 
+    public HashJoinNode(PlanNodeId id, PlanNode outer, PlanNode inner, JoinOperator joinOp, List<Expr> eqJoinConjuncts,
```
Why this PR include `planner`?

Kikyou1997 replied:
> Why this PR include `planner`?

And any other modification to Planner is based on the similiar reason as commented above

jackwener commented on fe/fe-core/src/main/java/org/apache/doris/nereids/properties/PhysicalProperties.java:
```diff
@@ -17,9 +17,20 @@
 
 package org.apache.doris.nereids.properties;
 
+import org.apache.doris.analysis.DistributionDesc;
```
config `style` according to [doc](https://doris.apache.org/zh-CN/developer-guide/java-format-code.html#import-order)

Kikyou1997 replied:
done

Kikyou1997 committed:
remove useless import

Kikyou1997 committed:
fix code style

Kikyou1997 committed:
fix code style

924060929 commented on fe/fe-core/src/main/java/org/apache/doris/nereids/operators/AbstractOperator.java:
```diff
@@ -17,13 +17,18 @@
  */
 public abstract class AbstractOperator<TYPE extends AbstractOperator<TYPE>> implements Operator<TYPE> {
     protected final OperatorType type;
+    protected long limited;
```
```suggestion
    protected final long limited;
```

Kikyou1997 replied:
done

924060929 commented on fe/fe-core/src/main/java/org/apache/doris/nereids/PlanOperatorVisitor.java:
```diff
@@ -0,0 +1,57 @@
+public abstract class PlanOperatorVisitor<R, C> {
+    public abstract R visit(PhysicalPlan<? extends PhysicalPlan, ? extends PhysicalOperator> physicalPlan, C context);
+
+    public R visitPhysicalAggregationPlan(PhysicalPlan<? extends PhysicalPlan, ? extends PhysicalOperator> physicalPlan,
```
```suggestion
    public R visitPhysicalAggregation(PhysicalPlan<? extends PhysicalPlan, PhysicalAggregation> aggregationPlan, C context);
```

Kikyou1997 replied:
done

924060929 commented on fe/fe-core/src/main/java/org/apache/doris/nereids/operators/AbstractOperator.java:
```diff
@@ -33,4 +38,17 @@ public AbstractOperator(OperatorType type) {
+        return limited;
+    }
+
+    public void setLimited(long limited) {
```
Operator and Plan's properties should be immutable. So no setter.

Kikyou1997 replied:
done

924060929 commented on fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalAggregation.java:
```diff
@@ -0,0 +1,100 @@
+
+public class PhysicalAggregation extends PhysicalUnaryOperator<PhysicalAggregation, PhysicalPlan> {
+
+    private List<Expression> groupByExprList;
```
operator's properties should be final

Kikyou1997 replied:
done

924060929 commented on fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalHashJoin.java:
```diff
@@ -0,0 +1,49 @@
+
+public class PhysicalHashJoin extends PhysicalBinaryOperator<PhysicalHashJoin, PhysicalPlan, PhysicalPlan> {
+
+    private JoinType joinType;
```
ditto

Kikyou1997 replied:
done

924060929 commented on fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalSort.java:
```diff
@@ -0,0 +1,67 @@
+
+public class PhysicalSort extends PhysicalUnaryOperator<PhysicalSort, PhysicalPlan> {
+
+    private int offset;
```
ditto

Kikyou1997 replied:
done

924060929 commented on fe/fe-core/src/main/java/org/apache/doris/nereids/PlanOperatorVisitor.java:
```diff
@@ -0,0 +1,57 @@
+
+@SuppressWarnings("rawtypes")
+public abstract class PlanOperatorVisitor<R, C> {
+    public abstract R visit(PhysicalPlan<? extends PhysicalPlan, ? extends PhysicalOperator> physicalPlan, C context);
```
PhysicalPlan should be Plan.
```
public abstract R visit(Plan<? extends Plan, ? extends PlanOperator> plan, C context);
```

Kikyou1997 replied:
done

wangshuo128 commented on fe/fe-core/src/main/java/org/apache/doris/nereids/PlanOperatorVisitor.java:
```diff
@@ -0,0 +1,57 @@
+import org.apache.doris.nereids.trees.plans.physical.PhysicalPlan;
+
+@SuppressWarnings("rawtypes")
+public abstract class PlanOperatorVisitor<R, C> {
```
Is this class a generic visitor for all the operators, i.e., both logical and physical, or it's a dedicated visitor only for physical operators?

Kikyou1997 replied:
It's a generic visitor

Kikyou1997 committed:
1. set properties of physical nodes as final
2. modify the parameter of base visit method
3. rename the visitxxxNode parameter name to concrete plan
4. implement the accept method in the concrete physical node
5. modify the plan type of the parameter of Operator to base Plan

Kikyou1997 committed:
add concrete operator type for PlanVisitor's method

Kikyou1997 committed:
add some comments

Kikyou1997 committed:
fix check style

wangshuo128 commented on fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java:
```diff
@@ -38,49 +38,48 @@
+ * <p>
  * The tree structure looks as follows:
  * - for non-distinct aggregation:
- *   - aggInfo: contains the original aggregation functions and grouping exprs
```
We'd better not change the original comment indents.
You could wrap the comments with the `<pre></pre>` HTML tag, thus the indents wouldn't be modified when formatting the code style.

Kikyou1997 replied:
done

wangshuo128 commented on fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java:
```diff
@@ -38,49 +38,48 @@
-        FIRST_MERGE,
-        SECOND,
-        SECOND_MERGE;
+        FIRST, FIRST_MERGE, SECOND, SECOND_MERGE;
```
using the original code style?

Kikyou1997 replied:
done

wangshuo128 commented on fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java:
```diff
@@ -277,29 +271,27 @@ public static boolean estimateIfContainsMultiDistinct(List<FunctionCallExpr> dis
-     *
+     * of DISTINCT aggregate functions
+     * <p>
      * Example:
```
using the original indents?

Kikyou1997 replied:
done

wangshuo128 commented on fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java:
```diff
@@ -414,20 +405,20 @@ public void getRefdSlots(List<SlotId> ids) {
      * Substitute all the expressions (grouping expr, aggregate expr) and update our
      * substitution map according to the given substitution map:
      * - smap typically maps from tuple t1 to tuple t2 (example: the smap of an
-     *   inline view maps the virtual table ref t1 into a base table ref t2)
```
ditto.

wangshuo128 commented on fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java:
```diff
@@ -539,15 +526,14 @@ private Expr createCountDistinctAggExprParam(int firstIdx, int lastIdx,
      * - 'this' is the phase 1 aggregation
      * - grouping exprs are those of the original query (param origGroupingExprs)
      * - aggregate exprs for the DISTINCT agg fns: these are aggregating the grouping
-     *   slots that were added to the original grouping slots in phase 1;
```
ditto.

wangshuo128 commented on fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java:
```diff
@@ -733,8 +707,8 @@ public void createSmaps(Analyzer analyzer) {
      * - The parameters of the sum function may involve the columns of a materialized view.
      * - The type of this column may happen to be inconsistent with the column type of the base table.
      * - In order to ensure the correctness of the result,
-     *   the parameter type needs to be changed to the type of the materialized view column
```
ditto.

wangshuo128 commented on fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java:
```diff
@@ -761,10 +735,10 @@ public void updateTypeOfAggregateExprs() {
      * Mark slots required for this aggregation as materialized:
      * - all grouping output slots as well as grouping exprs
      * - for non-distinct aggregation: the aggregate exprs of materialized aggregate slots;
-     *   this assumes that the output slots corresponding to aggregate exprs have already
```
ditto.

Kikyou1997 committed:
revert unnecessary changes to comments

Kikyou1997 committed:
add comment to the modifications to old optimizer

Kikyou1997 committed:
fix code style

Kikyou1997 committed:
Merge branch 'master' of github.com:apache/incubator-doris into opt

Kikyou1997 committed:
remove redundant code

Kikyou1997 committed:
add some comment

Kikyou1997 committed:
add some comment

EmmyMiao87 reviewed (approved):
LGTM
